### PR TITLE
Upgrade OE SDK to use mbedTLS 2.16.2

### LINKED ISSUE #1457: Evaluate moving to mbedTLS v2.16 branch
mbedTLS has designated its latest update [v2.16.1](https://tls.mbed.org/download/start/mbedtls-2.16.1-apache.tgz) as of 2018-12-21 the new long-term stable production branch. We should evaluate the impact of moving to this new release:

- [ ] Reconcile breaking changes in the API surface with OE SDK use
- [ ] Evaluate whether this poses issues for link-time mbedTLS replacement for projects using mbedTLS updated with Everest TLS crypto.
- [ ] Evaluate cost-benefit of moving to 2.16.x
- [x] Determine if 2.7 branch is being EOL and timeframe for that.
- [ ] Request OSS SDL security review of new mbedTLS version.

### PR DESCRIPTION
**Pull request by CodeMonkeyLeet**
This PR consolidates and updates #2043 to target the next minor version of 2.16 LTS lineage. It ports the following changes:

- Update CHANGELOG.md to reflect new version of mbedTLS.
- Update 3rdparty/mbedtls/update.make to target v2.16.2.
    - This version no longer has a yotta/.gitignore that needs to be removed.
- Update mbed tests to use Python instead of Perl to build tests.
    - MbedTLS 2.16 no longer has the Perl script for building its tests.
    - Add Python dependency to install-windows-prereqs.ps1
- Remove redundant search for data file to pass to mbed tests.
- Add LSEEK syscall test hook required by updated mbed test suite.
- Add platform_util.c to host and mbedtls build sources.
    - This provides symbols referenced by other mbedTLS source files.
- Apply patch to x509write_crt.c for attested TLS support.

Fixes #1457.

### CODE REVIEW

**Comment by CodeMonkeyLeet** [comment]
@johnkord, your comment from #2043 is the only remaining blocker for the original PR:

> @EmilAlexandruStoica Looks like Windows fails with mbed tests because we reference some python scripts, and we don't install python as a dependency on Windows!

The changeset already has modifications to install Python as a dependency on Windows, but there's no automated system to redeploy dependencies on Windows machines today. All that needs to be done is to install Python on the CI/CD servers.

**Comment by johnkord** [comment]
@CodeMonkeyLeet Thanks Simon! Yep, we just need python on the Windows systems. @EmilAlexandruStoica has been working on that, I believe. Any updates on this @EmilAlexandruStoica ?

File: 3rdparty/mbedtls/mbedtls/library/x509write_crt.c [hunk 1]
```diff
@@ -325,25 +325,25 @@ static int x509_write_time( unsigned char **p, unsigned char *start,
...
     size_t sub_len = 0, pub_len = 0, sig_and_oid_len = 0, sig_len;
     size_t len = 0;
     mbedtls_pk_type_t pk_alg;
 
     /*
-     * Prepare data to be signed in tmp_buf
+     * Prepare data to be signed at the end of the target buffer
      */
-    c = tmp_buf + sizeof( tmp_buf );
+    c = buf + size;
```
Comment (anakrish) [inline 3rdparty/mbedtls/mbedtls/library/x509write_crt.c]: Is there plans to upstream this change?

File: CHANGELOG.md [hunk 2]
```diff
@@ -36,6 +36,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
          should enable it themselves after assessing its startup impact.
 - Removed support for the previously deprecated `OE_API_VERSION=1` APIs.
 - Update MUSL libc to version 1.1.21.
+- Update mbedTLS to version 2.16.2.
```
Comment (anakrish) [inline CHANGELOG.md]: Shouldn't we create a new v0.7 changed section instead of editing the section for v0.6?

File: tests/mbed/enc/enc.c [hunk 3]
```diff
@@ -182,9 +185,8 @@ int test(
     }
     else
     {
-        static const char* argv[] = {"test", "-v", "NULL"};
+        static const char* argv[] = {"test", "-v"};
```
Comment (anakrish) [inline tests/mbed/enc/enc.c]: Can you explain the change to remove the test name?

File: CHANGELOG.md (hunk 2 above)
Comment (CodeMonkeyLeet) [reply CHANGELOG.md]: Thanks, I messed up a merge with master, it should have replaced the 2.7.11 line in the Unreleased section. Fixed.

File: 3rdparty/mbedtls/mbedtls/library/x509write_crt.c (hunk 1 above)
Comment (CodeMonkeyLeet) [reply 3rdparty/mbedtls/mbedtls/library/x509write_crt.c]: Not to my knowledge, so I filed #2179 when I came across this during the update to 2.7.11

File: tests/mbed/enc/enc.c (hunk 3 above)
Comment (CodeMonkeyLeet) [reply tests/mbed/enc/enc.c]: Credit to @EmilAlexandruStoica for this work. It's just clean up to remove a completely unused parameter; there used to be extra work to find and path the data file, but it didn't actually change the set of tests that were run. I'll update the commit message to reflect this.

File: 3rdparty/mbedtls/mbedtls/library/x509write_crt.c (hunk 1 above)
Comment (johnkord) [reply 3rdparty/mbedtls/mbedtls/library/x509write_crt.c]: There was an issue opened by @soccerGB with mbedTLS several months ago about this: https://github.com/ARMmbed/mbedtls/issues/2631

From the comments, it looks like they've merged the changes in successfully, although I'm not sure which release they're available in: https://github.com/ARMmbed/mbedtls/pull/2632

**Comment by EmilAlexandruStoica** [comment]
> @CodeMonkeyLeet Thanks Simon! Yep, we just need python on the Windows systems. @EmilAlexandruStoica has been working on that, I believe. Any updates on this @EmilAlexandruStoica ?

I've connected to oe-bastion and forwarded the port 3389 so I can connect from the localhost to the Windows VM's. Then I tried to connect to the Windows VM's to install Python, but it seems that the RDP connection is not enabled.
Capture

File: 3rdparty/mbedtls/mbedtls/.travis.yml [hunk 4]
```diff
@@ -35,7 +42,7 @@ addons:
   coverity_scan:
     project:
       name: "ARMmbed/mbedtls"
-    notification_email: p.j.bakker@polarssl.org
+    notification_email: simon.butcher@arm.com
```
Comment (BRMcLaren) [inline 3rdparty/mbedtls/mbedtls/.travis.yml]: I'm curious if there is a better solution than a hard coded individual e-mail address, perhaps an org? 

The rationale being if Simon leaves the company, can we guarantee the correct person is notified or that their e-mail would continue to be monitored? Thus creating further changes needed.

It seems this has happened before.

Comment (CodeMonkeyLeet) [reply 3rdparty/mbedtls/mbedtls/.travis.yml]: That's not part of the OE SDK project (it's just a clone of the ARM mbedTLS repo) so I wouldn't worry about it. If/when we move to a git submodule approach for dependencies, then this noise will go away, but in the meantime, you can ignore 72f5a736 as being just a straight clone of the 2.16.2 mbedTLS sources.

Comment (BRMcLaren) [reply 3rdparty/mbedtls/mbedtls/.travis.yml]: Sounds good, appreciate the clarification!

**Comment by johnkord** [comment]
I've updated the systems with Python 3.7.4 and ran the mbed tests manually on one of the systems and all the tests passed. Let's see what Jenkins thinks :)

bors try

**Comment by johnkord** [comment]
There appear to be two errors, not sure if they're related.

1. Building ELF enclaves on Linux first and then copying them to Windows for testing seems to have a problem with these tests because there appear to be some hard coded path embedded in the enclave:
```
test 13
        Start  13: tests/mbed/asn1write

13: Test command: C:\Jenkins\workspace\Bors_trying\build\tests\mbed\host\libmbedtest_host.exe "C:/Jenkins/workspace/Bors_trying/build/tests/mbed/host/mbedtest_suite_asn1write"
13: Test timeout computed to be: 480
13: RUNNING: asn1write
13: Failed to open test file: /home/jenkins/work_dir/workspace/Bors_trying@3/build/tests/mbed/enc/test_suite_asn1write.datax
13: Test failed: ..\tests\mbed\host\host.c(26): Test args.total > 0
 13/109 Test  #13: tests/mbed/asn1write ..................................***Exception: Exit code 0xc0000409
  0.15 sec
```

Notice the "Failed to open test file: /home/jenkins/work_dir/workspace/Bors_trying@3/build/tests/mbed/enc/test_suite_asn1write.datax" line.  Associated logs: https://oe-jenkins.eastus.cloudapp.azure.com/blue/rest/organizations/jenkins/pipelines/Bors/branches/trying/runs/995/nodes/64/steps/1042/log/?start=0

2. report_enc was not able to be copied for one of the stages from the linuxbin directory. I'm not sure why, but here's the output:

```
[218/427] cmd.exe /C "cd /D C:\Jenkins\workspace\Bors_trying\tests\report && "C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -E copy C:/Jenkins/workspace/Bors_trying\linuxbin\tests/report/enc/report_enc C:/Jenkins/workspace/Bors_trying/build/tests/report/host/report_enc"
FAILED: tests/report/tests/report_windows_include 
cmd.exe /C "cd /D C:\Jenkins\workspace\Bors_trying\tests\report && "C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -E copy C:/Jenkins/workspace/Bors_trying\linuxbin\tests/report/enc/report_enc C:/Jenkins/workspace/Bors_trying/build/tests/report/host/report_enc"
Error copying file "C:/Jenkins/workspace/Bors_trying\linuxbin\tests/report/enc/report_enc" to "C:/Jenkins/workspace/Bors_trying/build/tests/report/host/report_enc".
```
Associated logs: https://oe-jenkins.eastus.cloudapp.azure.com/blue/rest/organizations/jenkins/pipelines/Bors/branches/trying/runs/995/nodes/64/steps/1042/log/?start=0

**Comment by CodeMonkeyLeet** [comment]
@johnkord Much appreciated! 
- The first looks like a regression introduced by the removal of the explicit test data path specification from the mbed tests, so I'll restore that and retry. 
- The latter looks like a sporadic failure, given that the other similar configs that follow don't hit the same issue; something to keep track of in terms of the CI/CD robustness perhaps.

**Pull request merged**